### PR TITLE
Api 30 : add partial update of a category in the API

### PR DESCRIPTION
**Pull request by ahocquard**
[//]: <> (<3 Thanks for taking the time to contribute! You're awesome! <3)

[//]: <> (If you've never contributed to this repository before, please read https://github.com/akeneo/pim-community-dev/blob/master/.github/CONTRIBUTING.md)

**Description (for Contributor and Core Developer)**

Method PATCH to update partially a category. 
It's an upsert operation : if the category does not exist, it creates it.
The code is optional in the body (already provided in the url).

In case of a creation, the code should be the same between the url query parameter and the body (if it's provided in the body).

[//]: <> (What does this Pull Request do? reference the related issue?)

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | -
| Added Behats                      | -
| Added integration tests           | OK
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Bundle/ApiBundle/Controller/Rest/CategoryController.php [hunk 1]
```diff
@@ -209,4 +236,69 @@ protected function updateCategory(CategoryInterface $category, $data)
...
+    }
+
+    /**
+     * Get a response with HTTP code 204 when an object is updated.
+     *
+     * @return Response
+     */
+    protected function getUpdateResponse()
+    {
+        $response = new Response(null, Response::HTTP_NO_CONTENT);
```
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Controller/Rest/CategoryController.php]: add a header Location

File: src/Pim/Bundle/ApiBundle/Controller/Rest/CategoryController.php [hunk 2]
```diff
@@ -209,4 +236,69 @@ protected function updateCategory(CategoryInterface $category, $data)
...
+     * @param string $code code provided in the url
+     * @param array  $data code provided in the request body
+     *
+     * @throws UnprocessableEntityHttpException
+     */
+    protected function validateCodeConsistency($code, array $data)
+    {
+        if (isset($data['code']) && $code !== $data['code']) {
+            throw new UnprocessableEntityHttpException(
+                'The code provided in the request body must match the code provided in the url.'
```
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Controller/Rest/CategoryController.php]: maybe inject in the message $data['code'] & $code

File: src/Pim/Bundle/ApiBundle/Controller/Rest/CategoryController.php [hunk 3]
```diff
@@ -209,4 +236,69 @@ protected function updateCategory(CategoryInterface $category, $data)
...
+     * @return Response
+     */
+    protected function getUpdateResponse()
+    {
+        $response = new Response(null, Response::HTTP_NO_CONTENT);
+
+        return $response;
+    }
+
+
```
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Controller/Rest/CategoryController.php]: remove

File: src/Pim/Bundle/ApiBundle/Controller/Rest/CategoryController.php (hunk 1 above)
Comment (ahocquard) [inline src/Pim/Bundle/ApiBundle/Controller/Rest/CategoryController.php]: if it's update, it's useless. It's only in case of a creation.

**Pull request merged**